"""
Agent Stats Integration Test

æµ‹è¯•Agentç»Ÿè®¡ç³»ç»Ÿçš„å®Œæ•´æµç¨‹ï¼š
1. è®°å½•Agentæ‰§è¡Œ
2. æŸ¥è¯¢ç»Ÿè®¡æ•°æ®
3. éªŒè¯æ•°æ®å‡†ç¡®æ€§
"""
import asyncio
from datetime import datetime, timedelta
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession
from sqlalchemy.orm import sessionmaker
import sys
import os

# Add backend to path
sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))

from app.models.agent_stats import AgentExecutionStats, Base
from app.services.agent_stats_service import AgentStatsService


async def test_agent_stats_integration():
    """å®Œæ•´é›†æˆæµ‹è¯•"""

    # åˆ›å»ºå†…å­˜æ•°æ®åº“ç”¨äºæµ‹è¯•
    engine = create_async_engine(
        "sqlite+aiosqlite:///:memory:",
        echo=False
    )

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    async_session = sessionmaker(
        engine,
        class_=AsyncSession,
        expire_on_commit=False
    )

    async with async_session() as session:
        service = AgentStatsService(session)

        print("ğŸ§ª å¼€å§‹Agentç»Ÿè®¡ç³»ç»Ÿé›†æˆæµ‹è¯•\n")

        # æµ‹è¯•æ•°æ®
        test_user_id = 1
        test_session_id = "test_session_123"
        test_request_id = "req_456"

        # 1. æµ‹è¯•è®°å½•Agentæ‰§è¡Œ
        print("1ï¸âƒ£ æµ‹è¯•è®°å½•Agentæ‰§è¡Œ...")

        test_cases = [
            {
                "agent_type": "knowledge",
                "tool_name": "knowledge_search",
                "operation": "æ£€ç´¢çŸ¥è¯†åº“",
                "duration": 450,
                "status": "success"
            },
            {
                "agent_type": "math",
                "tool_name": "calculate_math",
                "operation": "è®¡ç®—å…¬å¼",
                "duration": 320,
                "status": "success"
            },
            {
                "agent_type": "code",
                "tool_name": "run_code",
                "operation": "æ‰§è¡ŒPythonä»£ç ",
                "duration": 680,
                "status": "success"
            },
            {
                "agent_type": "data_analysis",
                "tool_name": "analyze_data",
                "operation": "åˆ†ææ•°æ®é›†",
                "duration": 890,
                "status": "success"
            },
            {
                "agent_type": "translation",
                "tool_name": "translate_text",
                "operation": "ç¿»è¯‘æ–‡æœ¬",
                "duration": 230,
                "status": "success"
            },
            {
                "agent_type": "knowledge",
                "tool_name": "knowledge_search",
                "operation": "æ£€ç´¢çŸ¥è¯†åº“",
                "duration": 520,
                "status": "failed"
            },
        ]

        for i, case in enumerate(test_cases):
            start_time = datetime.utcnow() - timedelta(seconds=case['duration'] / 1000)
            end_time = start_time + timedelta(milliseconds=case['duration'])

            await service.record_agent_execution(
                user_id=test_user_id,
                session_id=test_session_id,
                request_id=f"{test_request_id}_{i}",
                agent_type=case['agent_type'],
                started_at=start_time,
                completed_at=end_time,
                status=case['status'],
                tool_name=case['tool_name'],
                operation=case['operation']
            )
            print(f"   âœ… è®°å½•: {case['agent_type']} - {case['operation']} ({case['duration']}ms)")

        # 2. æµ‹è¯•è·å–ç”¨æˆ·ç»Ÿè®¡æ¦‚è§ˆ
        print("\n2ï¸âƒ£ æµ‹è¯•è·å–ç”¨æˆ·ç»Ÿè®¡æ¦‚è§ˆ...")
        overview = await service.get_user_stats(test_user_id, days=30)

        print(f"   ğŸ“Š æ€»ä½“ç»Ÿè®¡:")
        print(f"      - æ€»æ‰§è¡Œæ¬¡æ•°: {overview['overall']['total_executions']}")
        print(f"      - å¹³å‡è€—æ—¶: {overview['overall']['avg_duration_ms']}ms")
        print(f"      - æ€»ä¼šè¯æ•°: {overview['overall']['total_sessions']}")

        print(f"   ğŸ“ˆ æŒ‰Agentç»Ÿè®¡:")
        for agent in overview['by_agent']:
            print(f"      - {agent['agent_type']}: {agent['count']}æ¬¡, "
                  f"å¹³å‡{agent['avg_duration_ms']}ms, "
                  f"æˆåŠŸç‡{agent['success_rate']:.1f}%")

        # 3. æµ‹è¯•è·å–Top Agent
        print("\n3ï¸âƒ£ æµ‹è¯•è·å–Top Agent...")
        top_agents = await service.get_most_used_agents(test_user_id, limit=3)

        for i, agent in enumerate(top_agents, 1):
            print(f"   ğŸ† Top {i}: {agent['agent_name']} - {agent['usage_count']}æ¬¡")

        # 4. æµ‹è¯•è·å–æ€§èƒ½æŒ‡æ ‡
        print("\n4ï¸âƒ£ æµ‹è¯•è·å–æ€§èƒ½æŒ‡æ ‡...")
        metrics = await service.get_performance_metrics(
            user_id=test_user_id,
            days=30
        )

        print(f"   âš¡ æ€§èƒ½æŒ‡æ ‡:")
        print(f"      - æ€»æ‰§è¡Œ: {metrics['total_executions']}æ¬¡")
        print(f"      - å¹³å‡è€—æ—¶: {metrics['avg_duration_ms']}ms")
        print(f"      - ä¸­ä½æ•°: {metrics['median_duration_ms']}ms")
        print(f"      - P95: {metrics['p95_duration_ms']}ms")
        print(f"      - æˆåŠŸç‡: {metrics['success_rate']:.1f}%")

        # 5. éªŒè¯æ•°æ®å‡†ç¡®æ€§
        print("\n5ï¸âƒ£ éªŒè¯æ•°æ®å‡†ç¡®æ€§...")

        # éªŒè¯æ€»æ‰§è¡Œæ¬¡æ•°
        expected_count = len(test_cases)
        actual_count = overview['overall']['total_executions']
        assert actual_count == expected_count, f"æœŸæœ›{expected_count}æ¬¡ï¼Œå®é™…{actual_count}æ¬¡"
        print(f"   âœ… æ‰§è¡Œæ¬¡æ•°éªŒè¯é€šè¿‡: {actual_count}æ¬¡")

        # éªŒè¯Agentç±»å‹æ•°é‡
        expected_agent_types = len(set(c['agent_type'] for c in test_cases))
        actual_agent_types = len(overview['by_agent'])
        assert actual_agent_types == expected_agent_types, f"æœŸæœ›{expected_agent_types}ç§Agentï¼Œå®é™…{actual_agent_types}ç§"
        print(f"   âœ… Agentç±»å‹æ•°é‡éªŒè¯é€šè¿‡: {actual_agent_types}ç§")

        # éªŒè¯æˆåŠŸç‡è®¡ç®—
        success_count = sum(1 for c in test_cases if c['status'] == 'success')
        expected_success_rate = (success_count / len(test_cases)) * 100
        actual_success_rate = metrics['success_rate']
        assert abs(actual_success_rate - expected_success_rate) < 0.1, f"æˆåŠŸç‡è®¡ç®—é”™è¯¯"
        print(f"   âœ… æˆåŠŸç‡è®¡ç®—éªŒè¯é€šè¿‡: {actual_success_rate:.1f}%")

        print("\nğŸ‰ æ‰€æœ‰æµ‹è¯•é€šè¿‡ï¼Agentç»Ÿè®¡ç³»ç»Ÿè¿è¡Œæ­£å¸¸ã€‚")

        # æ‰“å°æ±‡æ€»ä¿¡æ¯
        print("\nğŸ“Š æµ‹è¯•æ•°æ®æ±‡æ€»:")
        print(f"   - æµ‹è¯•ç”¨æˆ·ID: {test_user_id}")
        print(f"   - æµ‹è¯•ä¼šè¯æ•°: {overview['overall']['total_sessions']}")
        print(f"   - æ€»æ‰§è¡Œæ¬¡æ•°: {overview['overall']['total_executions']}")
        print(f"   - æ¶‰åŠAgentæ•°: {len(overview['by_agent'])}")
        print(f"   - å¹³å‡è€—æ—¶: {overview['overall']['avg_duration_ms']}ms")

        return True


async def test_agent_type_mapping():
    """æµ‹è¯•Agentç±»å‹æ˜ å°„"""
    print("\nğŸ§ª æµ‹è¯•Agentç±»å‹æ˜ å°„...")

    from app.orchestration.orchestrator import get_agent_type_for_tool
    from app.gen.agent.v1 import agent_service_pb2

    test_cases = [
        ('knowledge_search', agent_service_pb2.KNOWLEDGE),
        ('calculate_math', agent_service_pb2.MATH),
        ('run_code', agent_service_pb2.CODE),
        ('analyze_data', agent_service_pb2.DATA_ANALYSIS),
        ('translate_text', agent_service_pb2.TRANSLATION),
        ('generate_image', agent_service_pb2.IMAGE),
        ('process_audio', agent_service_pb2.AUDIO),
        ('write_content', agent_service_pb2.WRITING),
        ('solve_logic', agent_service_pb2.REASONING),
        ('create_task', agent_service_pb2.ORCHESTRATOR),
    ]

    for tool_name, expected in test_cases:
        actual = get_agent_type_for_tool(tool_name)
        assert actual == expected, f"å·¥å…·{tool_name}æ˜ å°„é”™è¯¯"
        print(f"   âœ… {tool_name} -> {agent_service_pb2.AgentType.Name(actual)}")

    print("   ğŸ‰ Agentç±»å‹æ˜ å°„æµ‹è¯•é€šè¿‡ï¼")
    return True


async def main():
    """è¿è¡Œæ‰€æœ‰æµ‹è¯•"""
    try:
        await test_agent_type_mapping()
        await test_agent_stats_integration()
        print("\n" + "="*60)
        print("âœ… æ‰€æœ‰é›†æˆæµ‹è¯•é€šè¿‡ï¼")
        print("="*60)
        return True
    except Exception as e:
        print(f"\nâŒ æµ‹è¯•å¤±è´¥: {e}")
        import traceback
        traceback.print_exc()
        return False


if __name__ == "__main__":
    result = asyncio.run(main())
    sys.exit(0 if result else 1)
